File: js-src/src/storage/Delta.ts
import { Change, ChangeData, ChangeType } from "../Change";
import { getLogger } from "../logging";

let log = getLogger('Delta')

export type DeltaData = {
    [key: string]: ChangeData;
}

export class Delta {
    // Max entity key depth: e.g. entity.content.image.height - 3 levels
    // 1 property scope, 2 property object, 3 key
    private _data: DeltaData = {};

    constructor(data: DeltaData) {
        this._data = data;
    }

    static fromChanges(changes: Change[]): Delta {
        let delta = new Delta({});
        for (const change of changes) {
            delta.addChangeFromData(change.data)
        }
        return delta;
    }

    get data(): DeltaData {
        return this._data;
    }

    setData(data: DeltaData) {
        this._data = data;
    }

    *changes(): Generator<Change> {
        for (const changeData of Object.values(this._data)) {
            yield new Change(changeData);
        }
    }

    copy() {
        return new Delta(structuredClone(this._data));
    }

    reversed(): Delta {
        // Makes the changes in place
        const reversedData: DeltaData = {};
        for (const [entityId, reverseComponent, forwardComponent] of Object.values(this._data)) {
            reversedData[entityId] = [entityId, forwardComponent, reverseComponent];
        }
        return new Delta(reversedData);
    }

    addChangeFromData(changeData: ChangeData) {
        const [entityId] = changeData;
        if (this._data[entityId] !== undefined) {
            // Merge the changes
            this.mergeChangeWithPriority(new Change(changeData));
        } else {
            this._data[entityId] = changeData;
        }
    }

    removeChange(entityId: string) {
        if (this._data[entityId] === undefined) {
            throw Error('Entity delta with that key not found');
        }
        delete this._data[entityId];
    }

    isEmpty(): boolean {
        for (const [_, oldState, newState] of Object.values(this._data)) {
            if (Object.keys(oldState).length > 0 || Object.keys(newState).length > 0) {
                return false;
            }
        }
        return true;
    }

    entityIds(): string[] {
        return Object.keys(this._data);
    }

    changeData(entityId: string): ChangeData | undefined {
        return this._data[entityId];
    }

    change(entityId: string): Change | undefined{
        const changeData = this.changeData(entityId);
        if (changeData === undefined) {
            return undefined;
        }
        return new Change(changeData);
    }
    mergeChangeWithPriority(change: Change) {
        /**
         * Merges a new change (second/next) into the existing delta (first)
         * for a specific entity, as if it was applied afterward.
         *
         * This function enforces a priority or ordering, meaning
         * "the second/next change overrides or merges into the first."
         *
         * Supported patterns for merging:
         *  1) Update > Update   : Combine forward/reverse fields
         *  2) Create > Update   : Extend the 'create' forward component
         *  3) Delete > Create   : Treat as an 'update' (old data => new data)
         *  4) Create > Delete   : The creation is effectively canceled
         *  5) Otherwise         : Log as an error or do nothing
         *
         * Notes:
         *  - 'forwardComponent' describes how to go from the "old" to "new" state.
         *  - 'reverseComponent' describes how to go back from the "new" to "old" state.
         */
        const firstChange = this.change(change.entityId);

        // If no existing (first) change for this entity, just add the new one
        if (!firstChange) {
            this.addChangeFromData(change.data);
            return;
        }

        // Identify what kind of changes we're merging:
        const firstCT = firstChange.type();  // e.g. CREATE, UPDATE, DELETE
        const nextCT = change.type();        // the new change being merged in

        /**
         * 1) UPDATE > UPDATE
         *    We combine the forward components to reflect the sum of both updates,
         *    and merge reverse components to preserve the ability to revert back
         *    to the earliest "old" state if necessary.
         */
        if (firstCT === ChangeType.UPDATE && nextCT === ChangeType.UPDATE) {
            const firstForward = { ...firstChange.forwardComponent };
            const firstReverse = { ...firstChange.reverseComponent };
            const nextForward = change.forwardComponent;
            const nextReverse = change.reverseComponent;

            // Forward props: new update fields appended on top
            if (nextForward) {
                firstChange.forwardComponent = { ...firstForward, ...nextForward };
            }
            // Reverse props: nextReverse overrides firstReverse to revert properly
            if (nextReverse) {
                firstChange.reverseComponent = { ...nextReverse, ...firstReverse };
            }

        /**
         * 2) CREATE > UPDATE
         *    We interpret the second update as simply extending the forward part
         *    of the original CREATE. Because the entity was just created, we can
         *    "stack" these updates into its forward component.
         */
        } else if (firstCT === ChangeType.CREATE && nextCT === ChangeType.UPDATE) {
            if (change.forwardComponent) {
                firstChange.forwardComponent = {
                    ...firstChange.forwardComponent,
                    ...change.forwardComponent
                };
            }

        /**
         * 3) DELETE > CREATE
         *    The "delete" said "this entity used to exist, now it doesn't."
         *    The "create" says "here's a new version of the entity."
         *    Interpreting them together means we effectively have an 'update' from
         *    the old data to the newly created data (old => new).
         */
        } else if (firstCT === ChangeType.DELETE && nextCT === ChangeType.CREATE) {
            const oldData = firstChange.reverseComponent;  // Pre-delete data
            const newData = change.forwardComponent;       // Newly created data
            this._data[change.entityId] = [
                change.entityId,
                oldData,    // reverse => revert to "deleted" or old state
                newData     // forward => adopt the new data
            ];

        /**
         * 4) CREATE > DELETE
         *    Means we created an entity and are now deleting it before any usage.
         *    Net effect = remove both changes; there's no entity at all.
         */
        } else if (firstCT === ChangeType.CREATE && nextCT === ChangeType.DELETE) {
            this.removeChange(change.entityId);

        /**
         * 5) EMPTY or unhandled
         *    If the new change is empty, do nothing; if it's an irrational
         *    sequence (like 'Update > Create' on an entity that didn't exist),
         *    log an error.
         */
        } else if (nextCT === ChangeType.EMPTY) {
            // Nothing to do for empty changes
        } else {
            log.error(
                'Irrational or unhandled delta sequence:',
                firstChange,
                '->',
                change
            );
        }

    }

    changeType(entityId: string, reverse: boolean = false): ChangeType {
        let change = this.change(entityId);
        if (!change) {
            throw Error('No entity delta for the given entityId');
        }
        if (reverse) {
            change = change.reversed();
        }
        return change.type();
    }

    mergeWithPriority(next: Delta) { // In place
        /**
         * Union with another delta up to the level of entity.key. I.e. if
         * there's an update for an entity - if the updated key in both deltas
         * is the same - the second one ('other') will override the key in
         * the forward operation. And the first one (this) will override in
         * the backward direction
         */
        // Iterate over all entity deltas in the next delta
        for (const change of next.changes()) {
            this.mergeChangeWithPriority(change);
        }
    }
}

export function squishDeltas(deltas: DeltaData[]) {
    const squishedDelta = new Delta({});

    for (let delta of deltas) {
        squishedDelta.mergeWithPriority(new Delta(delta));
    }

    return squishedDelta;
}

function changeTypeFromData(
    entityDelta: ChangeData,
    reverse: boolean = false
): ChangeType {
    const [_, reverseDelta, forwardDelta] = entityDelta;
    const [componentToCheck, otherComponent] = reverse ? [forwardDelta, reverseDelta] : [reverseDelta, forwardDelta];

    const forwardPresent = Object.keys(otherComponent).length > 0;
    const reversePresent = Object.keys(componentToCheck).length > 0;

    if (forwardPresent && reversePresent) {
        return ChangeType.UPDATE;
    } else if (forwardPresent) {
        return ChangeType.CREATE;
    } else if (reversePresent) {
        return ChangeType.DELETE;
    } else {
        return ChangeType.EMPTY;
    }
}

File: py-src/fusion/libs/entity/delta.py
from __future__ import annotations

from copy import deepcopy
from typing import Any, Generator

from fusion import get_logger
from fusion.libs.entity.change import Change, ChangeData, ChangeTypes

log = get_logger(__name__)


# DeltaData is a dict of entity_id -> ChangeData
DeltaData = dict[str, ChangeData]


def _entity_keys_are_equal(value1: Any, value2: Any, depth: int = 1) -> bool:
    """Deep equality check for entity property values, up to 3 levels.

    Max depth matches entity property nesting convention:
    e.g. entity.content.image.height = 3 levels
    (1: property scope, 2: property object, 3: leaf key)
    """
    if depth > 3:
        raise ValueError(
            "Entity key comparison exceeded max depth of 3 — "
            "entity properties should not nest deeper than "
            "property.object.key (e.g. content.image.height)."
        )

    if type(value1) != type(value2):
        return False

    if value1 is None or value2 is None:
        return value1 == value2

    if not isinstance(value1, (dict, list)):
        return value1 == value2

    if isinstance(value1, list):
        if len(value1) != len(value2):
            return False
        return all(
            _entity_keys_are_equal(v1, v2, depth + 1) for v1, v2 in zip(value1, value2)
        )

    # Both are dicts
    if value1.keys() != value2.keys():
        return False
    return all(_entity_keys_are_equal(value1[k], value2[k], depth + 1) for k in value1)


def _change_reduces_to_none(change: Change) -> bool:
    """Check if an update change is a no-op (forward == reverse for all keys)."""
    if not change.is_update():
        return False

    reverse = dict(change.reverse_component)
    forward = dict(change.forward_component)

    for key in list(reverse.keys()):
        if key not in forward:
            continue
        if _entity_keys_are_equal(reverse[key], forward[key]):
            del reverse[key]
            del forward[key]

    change.reverse_component = reverse
    change.forward_component = forward

    return len(reverse) == 0 and len(forward) == 0


class Delta:
    """A collection of Changes grouped by entity ID, with merge logic."""

    def __init__(self, data: DeltaData | None = None):
        self._data: DeltaData = data or {}

    @staticmethod
    def from_changes(changes: list[Change]) -> Delta:
        delta = Delta()
        for change in changes:
            delta.add_change_from_data(change.data)
        return delta

    @property
    def data(self) -> DeltaData:
        return self._data

    def changes(self) -> Generator[Change, None, None]:
        for change_data in self._data.values():
            yield Change(change_data)

    def copy(self) -> Delta:
        return Delta(deepcopy(self._data))

    def reversed(self) -> Delta:
        reversed_data: DeltaData = {}
        for entity_id, (eid, reverse, forward) in self._data.items():
            reversed_data[entity_id] = (eid, forward, reverse)
        return Delta(reversed_data)

    def add_change_from_data(self, change_data: ChangeData):
        entity_id = change_data[0]
        if entity_id in self._data:
            self.merge_change_with_priority(Change(change_data))
        else:
            self._data[entity_id] = change_data

    def remove_change(self, entity_id: str):
        if entity_id not in self._data:
            raise KeyError(f"No change for entity {entity_id}")
        del self._data[entity_id]

    def is_empty(self) -> bool:
        for _, reverse, forward in self._data.values():
            if len(reverse) > 0 or len(forward) > 0:
                return False
        return True

    def entity_ids(self) -> list[str]:
        return list(self._data.keys())

    def change_data(self, entity_id: str) -> ChangeData | None:
        return self._data.get(entity_id)

    def change(self, entity_id: str) -> Change | None:
        cd = self.change_data(entity_id)
        if cd is None:
            return None
        return Change(cd)

    def merge_change_with_priority(self, change: Change):
        """Merge a new change into the existing delta for the same entity.

        Supported merge patterns:
        1) Update > Update: combine forward/reverse fields
        2) Create > Update: extend the create's forward component
        3) Delete > Create: becomes an update (old → new)
        4) Create > Delete: cancels out (removed entirely)
        5) Otherwise: log error
        """
        first_change = self.change(change.entity_id)

        if first_change is None:
            self.add_change_from_data(change.data)
            return

        first_ct = first_change.type()
        next_ct = change.type()

        # 1) Update > Update
        if first_ct == ChangeTypes.UPDATE and next_ct == ChangeTypes.UPDATE:
            merged_forward = {
                **first_change.forward_component,
                **change.forward_component,
            }
            merged_reverse = {
                **change.reverse_component,
                **first_change.reverse_component,
            }
            first_change.forward_component = merged_forward
            first_change.reverse_component = merged_reverse
            if _change_reduces_to_none(first_change):
                self.remove_change(change.entity_id)
            else:
                self._data[change.entity_id] = first_change.data

        # 2) Create > Update
        elif first_ct == ChangeTypes.CREATE and next_ct == ChangeTypes.UPDATE:
            merged_forward = {
                **first_change.forward_component,
                **change.forward_component,
            }
            first_change.forward_component = merged_forward
            self._data[change.entity_id] = first_change.data

        # 3) Delete > Create
        elif first_ct == ChangeTypes.DELETE and next_ct == ChangeTypes.CREATE:
            old_data = first_change.reverse_component
            new_data = change.forward_component
            self._data[change.entity_id] = (change.entity_id, old_data, new_data)
            merged_change = self.change(change.entity_id)
            if _change_reduces_to_none(merged_change):
                self.remove_change(change.entity_id)
            else:
                self._data[change.entity_id] = merged_change.data

        # 4) Create > Delete
        elif first_ct == ChangeTypes.CREATE and next_ct == ChangeTypes.DELETE:
            self.remove_change(change.entity_id)

        # 5) Empty or irrational
        elif next_ct == ChangeTypes.EMPTY:
            pass
        else:
            log.error(
                f"[merge_change_with_priority] Irrational delta sequence. "
                f"Merging {change} INTO {first_change}"
            )

    def merge_with_priority(self, next_delta: Delta):
        """Merge another delta into this one, in place."""
        for change in next_delta.changes():
            self.merge_change_with_priority(change)


def squash_deltas(deltas: list[DeltaData]) -> Delta:
    """Reduce an array of DeltaData dicts into one Delta."""
    squashed = Delta()
    for delta_data in deltas:
        squashed.merge_with_priority(Delta(delta_data))
    return squashed
